crates/chia-consensus/src/merkle_set.rs

```diff
@@ -147,567 +185,892 @@ fn radix_sort(range: &mut [[u8; 32]], depth: u8) -> ([u8; 32], NodeType) {
+        assert_eq!(
+            rebuilt.hash_cache[rebuilt.hash_cache.len() - 1],
+            tree.hash_cache[tree.hash_cache.len() - 1]
+        );
+    }
+
+    // this test generates a 1000000 vecs filled with 500 random data hashes
+    // It then creates a proof for one of the leafs and deserializes the proof and compares it to the original
+    #[test]
+    fn test_random_bytes() {
+        for _i in [1..1000000] {
+            // Create a random number generator
+            let mut small_rng = SmallRng::from_entropy();
+
+            // Generate a random length for the Vec
+            let vec_length: usize = small_rng.gen_range(0..=500);
+
+            // Generate a Vec of random [u8; 32] arrays
+            let mut random_data: Vec<[u8; 32]> = Vec::with_capacity(vec_length);
+            for _ in 0..vec_length {
+                let mut array: [u8; 32] = [0; 32];
+                small_rng.fill(&mut array);
+                random_data.push(array);
+            }
+
+            let (root, tree) = generate_merkle_tree(&mut random_data);
+            assert_eq!(tree.hash_cache.len(), tree.nodes_vec.len());
+            assert_eq!(root, tree.hash_cache[tree.nodes_vec.len() - 1]);
+            assert_eq!(root, compute_merkle_set_root(&mut random_data));
+            let mut rng = rand::thread_rng();
+            let index = rng.gen_range(0..random_data.len());
+            let (included, proof) = tree.generate_proof(random_data[index]).unwrap();
+            assert!(included);
+            let rebuilt = deserialize_proof(&proof).unwrap();
+            assert_eq!(
+                rebuilt.hash_cache[rebuilt.hash_cache.len() - 1],
+                tree.hash_cache[tree.hash_cache.len() - 1]
+            );
+        }
+    }
+
+    #[test]
+    fn test_bad_proofs() {
+        for _i in [1..1000000] {
+            // Create a random number generator
+            let mut small_rng = SmallRng::from_entropy();
+
+            // Generate a random length for the Vec
+            let vec_length: usize = small_rng.gen_range(0..=500);
+
+            // Generate a Vec of random [u8; 32] arrays
+            let mut random_data: Vec<[u8; 32]> = Vec::with_capacity(vec_length);
+            for _ in 0..vec_length {
+                let mut array: [u8; 32] = [0; 32];
+                small_rng.fill(&mut array);
+                random_data.push(array);
+            }
+
+            let (root, tree) = generate_merkle_tree(&mut random_data);
+            assert_eq!(tree.hash_cache.len(), tree.nodes_vec.len());
+            assert_eq!(root, tree.hash_cache[tree.nodes_vec.len() - 1]);
+            assert_eq!(root, compute_merkle_set_root(&mut random_data));
+            let mut rng = rand::thread_rng();
+            let index = rng.gen_range(0..random_data.len());
+            let (included, proof) = tree.generate_proof(random_data[index]).unwrap();
+            assert!(included);
+            let rebuilt = deserialize_proof(&proof[0..proof.len() - 2]);
```

Comment:
these proofs are mostly valid, just truncated. I think a fuzzer that passes in truly arbitrary bytes would be very valuable.